### PR TITLE
Fix excessive duplication during concatenation of named vectors

### PR DESCRIPTION
**Pull request by lionel-**
* Add `vec_proxy_set_names()` which takes a `vctrs_owned` argument. This variant of `vec_set_names()` never dispatches and uses `vec_clone_referenced(x, owned)` instead of `r_clone_referenced()`.

```r
ints <- rep(list(set_names(1:3, letters[1:3])), 1e4)
bench::mark(
  base = unlist(ints),
  vctrs = vec_unchop(ints),
  check = FALSE
)[1:8]
#> # Before
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#> 1 base        669.7µs  843.4µs  1101.     351.66KB     7.82   563     4
#> 2 vctrs          1.5s     1.5s     0.667    2.24GB    29.3      1    44

#> # After
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#> 1 base       663.31µs 833.14µs     1203.     352KB    10.9    554     5
#> 2 vctrs        4.52ms   4.93ms      201.     625KB     4.18    96     2
```

* Don't repair row names on proxies. Instead, row names are now repaired as part of restoration of data frames.

```r
df <- new_data_frame(
  df_list(
    x = as.list(1:2),
    y = 1:2
  ),
  row.names = c("foo", "bar")
)
dfs <- rep(list(df), 3)
vec_unchop(dfs)

dfs <- rep(list(df), 1e2)
bench::mark(
  vctrs = vec_unchop(dfs),
  check = FALSE
)[1:8]
#> # Before
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#> 1 vctrs         449ms    635ms      1.57    85.4MB     1.57     2     2

#> # After
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#> 1 vctrs         905µs   1.04ms      948.    14.8KB     18.0   421     8
```

This issue isn't fixed in the case of df-cols because we proxy and restore these at each assignment. This particular bug is tracked in #1217 and will be solved once #1107 is resolved.

We are now tracking memory allocations with human readable sizes in golden tests. This should allow us to detect memory issues early.

This PR also imports the pairlist and attributes utils from the rlang C library. This is because I initially tried to solve the names duplication issue with `r_clone_shared()`. The attribute getters from the GNU R lib are problematic in that regard because they max-bump the namedness/refcount of attributes. I don't think this is appropriate behaviour because in most cases we are only inspecting the attribute. The rlang utils never bump refcounts, and they also never allocate (no need for protection). Even though these rlang utils are no longer necessary to fix the duplication issues since I switched to the `vctrs_owned` approach that Davis introduced, I kept them in the PR because they have better properties.

### CODE REVIEW

File: src/names.c [hunk 1]
```diff
@@ -742,19 +755,18 @@ SEXP vec_set_df_rownames(SEXP x, SEXP names) {
...
+SEXP vec_set_names_impl(SEXP x, SEXP names, bool proxy, const enum vctrs_owned owned) {
   check_names(x, names);
 
   if (is_data_frame(x)) {
-    return vec_set_df_rownames(x, names);
+    return vec_set_df_rownames(x, names, proxy, owned);
   }
 
   if (has_dim(x)) {
     return vec_set_rownames(x, names);
```
Comment (DavisVaughan) [inline src/names.c:766]: Does this need to know about `proxy`?

File: src/slice-assign.c [hunk 2]
```diff
@@ -378,25 +379,27 @@ static SEXP vec_assign_fallback(SEXP x, SEXP index, SEXP value) {
...
+  SEXP proxy_nms = PROTECT(vec_proxy_names(proxy));
   if (proxy_nms == R_NilValue) {
     proxy_nms = PROTECT(Rf_allocVector(STRSXP, vec_size(proxy)));
   } else {
-    proxy_nms = PROTECT(r_clone_referenced(proxy_nms));
+    proxy_nms = PROTECT(vec_clone_referenced(proxy_nms, owned));
   }
+  proxy_nms = PROTECT(chr_assign(proxy_nms, index, value_nms, owned));
 
-  proxy_nms = PROTECT(chr_assign(proxy_nms, index, value_nms, VCTRS_OWNED_true));
```
Comment (DavisVaughan) [inline src/slice-assign.c:396]: I always found it a bit odd that `VCTRS_OWNED_true` was constant here, so this is a nice change 👍

File: tests/testthat/test-c.R [hunk 3]
```diff
@@ -438,5 +438,47 @@ test_that("concatenation performs expected allocations", {
 
     "Doubles to integer"
     with_memory_prof(vec_unchop(dbls, ptype = int()))
+
+
+    "# Concatenation with with names"
```
Comment (DavisVaughan) [inline tests/testthat/test-c.R:443]:
```suggestion
    "# Concatenation with names"
```

File: src/names.c (hunk 1 above)
Comment (lionel-) [reply src/names.c]: Probably, now done!

**Pull request merged**